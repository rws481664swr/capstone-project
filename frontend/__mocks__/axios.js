import {BASE_URL} from "../src/config";

axios.get.mockImplementation((url) => {
    let result = url.slice(BASE_URL.length);
    if(result[result.length-1] === '/') result = result.slice(0, result.length-1);
    switch (result){
        case '/courses':
            return Promise.resolve({data: coursesList})
        case '/users':
            return Promise.resolve({data:usersList})
        case '/auth/token':
            return Promise.resolve({data:'token'})
    }
    return Promise.resolve({data: []})

})
const axios = {
    get: jest.fn().mockResolvedValue({
        data: {}
    }),
    post: jest.fn().mockResolvedValue({
        data: {}
    }),
    put: jest.fn().mockResolvedValue({
        data: {}
    }),
    delete: jest.fn().mockResolvedValue({
        data: {}
    })
}
export default axios