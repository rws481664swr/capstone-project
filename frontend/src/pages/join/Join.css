#join-course{
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: center;
    width: 100%;
    margin-bottom: 20px;
    font-size: 2rem;
    font-weight: 300;
    margin-top: 2rem;
}