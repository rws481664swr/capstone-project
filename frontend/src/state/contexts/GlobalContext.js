import {createContext, useContext} from "react";
import jwtDecode from "jwt-decode";
import useLocalStorageState from "../../hooks/useLocalStorageState";

const value = {
    username: undefined,
    role: undefined,
    _id: undefined,
    timestamp: undefined,
    token:null,
    setToken:()=>{}

}
const GlobalContext = createContext(value)
export default GlobalContext

export const GlobalContextProvider = ({children}) => {
    const [context,setContext] =
        useLocalStorageState( 'ctx',value||{})
    const setToken= (token)=>{
        if (token){
            const {

                username,
                role,
                _id,
                timestamp
            } = jwtDecode(token)
            setContext({username, role, _id, timestamp, token})
        }else{
            setContext({})
        }
    }
    const logout= ()=> {
        setToken(null)
    }
    return <GlobalContext.Provider value={{...context,setToken,logout}}>
        {children}
    </GlobalContext.Provider>
}
export const useGlobalContext = () =>     useContext(GlobalContext)

