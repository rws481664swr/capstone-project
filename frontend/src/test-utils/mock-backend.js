import {BASE_URL} from "../config";

export const extractKeysAsList= (obj)=>
    Object.keys(obj).map(key=>obj[key])
        .reduce((acc, val)=>[...acc,val],[])
export const MOCK_TOKEN = {current: 'TOKEN'}

export const spliceUrl = (url) => {
    const arr = BASE_URL.split('')
    const toReturn = url.split('')
    while (arr.length){
        arr.shift()
        toReturn.shift()
    }
    return toReturn.join('')

}