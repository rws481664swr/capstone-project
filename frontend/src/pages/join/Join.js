import './Join.css'
import LabeledInput from "../../components/General/LabeledInput/LabeledInput";
import Button from "../../components/General/Button/GenericButton/Button";
import useFlash from "../../hooks/form/useFlash";
import {useState} from "react";
import useJoinCourse from "./useJoinCourse";
import {useParams} from "react-router-dom";

const JoinCourseLabelText = 'Course ID'
const JoinCourseButtonText = 'Join'


const Join = () => {
    const {id:param}= useParams()
    const [id, setID]=useState(param?param:'')

    const  joinCourse= useJoinCourse(id)

    const [toRender, , , {danger}] = useFlash()

    const submit =  async (e5) => {
        e5.preventDefault()
        await joinCourse(id, danger)

    }



    return (
        <div>
            <h1>Join Course</h1>
            {toRender}
            <form className={'sr-container'} onSubmit={submit}>
                <LabeledInput
                    name={'course_id'}
                    value={id}
                    onChange={e => setID(e.target.value)}
                    label={JoinCourseLabelText}
                    id={'Join_course_id'}
                />
                <Button type={'submit'}>{JoinCourseButtonText}</Button>
            </form>
        </div>
    )

}
Join.LabelText = JoinCourseLabelText
Join.ButtonText = JoinCourseButtonText

export default Join;