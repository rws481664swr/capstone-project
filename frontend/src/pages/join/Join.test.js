import {act, fireEvent, render} from "@testing-library/react";
import Join from "./Join";
import GlobalContext from "../../state/contexts/GlobalContext";
import axios from 'axios'
import {MemoryRouter, useNavigate, useParams} from "react-router-dom";
import {MOCK_TOKEN} from "../../../mock-backend";

const defaultValue = {
    username: 'test',
    token: MOCK_TOKEN,
    timestamp: new Date()
}
const mockNavigate = jest.fn(() => ({id: null}))
jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: jest.fn(
        () => mockNavigate)
    , useParams: jest.fn()
}))
// const mockPost = jest.fn()
jest.mock('axios', () => ({
    ...jest.requireActual('axios'),
    post: jest.fn()
}))

const defaultId = '123'
const TestJoin = ({
                      value =
                          defaultValue,
                      id = defaultId,
                      error = ''
                  }) =>
    <MemoryRouter>
        <GlobalContext.Provider value={value}>
            <Join id={''} error={error}/>
        </GlobalContext.Provider>
    </MemoryRouter>

it('should render Join without params', () => {
    useParams.mockImplementationOnce(()=>({}))

    render(<TestJoin value={defaultValue}/>)
})
it('should render Join with params', () => {
    useParams.mockImplementationOnce(()=>({id: '123'}))

    render(<TestJoin value={defaultValue}/>)
})
it('should render Join snapshot without params', () => {
    // useParams.mockImplementationOnce(()=>({}))
    useParams.mockImplementationOnce(()=>({}))

    const {asFragment} = render(<TestJoin value={defaultValue}/>)
    expect(asFragment()).toMatchSnapshot()
})
it('should render Join snapshot with params', () => {
    // useParams.mockImplementationOnce(()=>({}))
    useParams.mockImplementationOnce(()=>({id: '123'}))

    const {asFragment} = render(<TestJoin value={defaultValue}/>)
    expect(asFragment()).toMatchSnapshot()
})
it('should join course via POST', async () => {
    useParams.mockImplementationOnce(()=>({id: '123'}))


    const navigateWasCalled = jest.fn()

    useNavigate.mockImplementation(() => (arg) => {
            navigateWasCalled(arg)
            return () => {
            }
        }
    )
    axios.post.mockImplementation((url) => {
        console.log(url)
        expect(url).toContain('courses/123/users/test')
        return Promise.resolve({data: {}})
    })
    // mockPost.mockResolvedValueOnce({data: {message:'error'}})

    // mockImplementation(() => Promise.resolve({data: {}}))

    const {getByLabelText, getByText} =
        await act(() => render(<TestJoin value={defaultValue}/>))
    const joinBox = getByLabelText(Join.LabelText)
    fireEvent.change(joinBox, {target: {value: '123'}})
    expect(joinBox.value).toBe('123')
    await act(() => fireEvent.click(getByText(Join.ButtonText)))
    expect(useNavigate).toHaveBeenCalled()
    expect(navigateWasCalled).toHaveBeenCalledWith('/courses/123')
})
it('should join course via POST form', async () => {
    useParams.mockImplementationOnce(()=>({id: '123'}))


    const navigateWasCalled = jest.fn()

    useNavigate.mockImplementation(() => (arg) => {
        navigateWasCalled(arg)
            return () => {
            }
        }
    )
    axios.post.mockImplementation((url) => {
        return Promise.resolve({data: {}})
    })

    // mockPost.mockImplementation(() => Promise.resolve({data: {}}))

    const {getByLabelText, getByText} =
        await act(() => render(<TestJoin value={defaultValue}/>))
    const joinBox = getByLabelText(Join.LabelText)
    fireEvent.change(joinBox, {target: {value: '123'}})
    expect(joinBox.value).toBe('123')
    await act(() => fireEvent.click(getByText(Join.ButtonText)))
    expect(useNavigate).toHaveBeenCalled()
    expect(navigateWasCalled).toHaveBeenCalledWith('/courses/123')
})
it('should load a course into params', async () => {
    useParams.mockImplementationOnce(() => ({id: '321'}))

    const {getByLabelText, getByText} =
        await act(() => render(<TestJoin value={defaultValue}/>))
    const joinBox = getByLabelText(Join.LabelText)
    expect(joinBox.value).toBe('321')


})

