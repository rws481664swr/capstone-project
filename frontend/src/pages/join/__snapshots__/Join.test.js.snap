// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Join snapshot with params 1`] = `
<DocumentFragment>
  <div>
    <h1>
      Join Course
    </h1>
    <form
      class="sr-container"
    >
      <div
        class="LabeledInput-div "
      >
        <label
          class="LabeledInput-input-label "
          for="Join_course_id"
        >
          Course ID
        </label>
        <div />
        <input
          class="LabeledInput-input-text "
          id="Join_course_id"
          name="course_id"
          type="text"
          value="123"
        />
      </div>
      <button
        class="undefined Button"
        type="submit"
      >
        Join
      </button>
    </form>
  </div>
</DocumentFragment>
`;

exports[`should render Join snapshot without params 1`] = `
<DocumentFragment>
  <div>
    <h1>
      Join Course
    </h1>
    <form
      class="sr-container"
    >
      <div
        class="LabeledInput-div "
      >
        <label
          class="LabeledInput-input-label "
          for="Join_course_id"
        >
          Course ID
        </label>
        <div />
        <input
          class="LabeledInput-input-text "
          id="Join_course_id"
          name="course_id"
          type="text"
          value=""
        />
      </div>
      <button
        class="undefined Button"
        type="submit"
      >
        Join
      </button>
    </form>
  </div>
</DocumentFragment>
`;
